src/validation.cpp

```diff
@@ -818,33 +814,40 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
                             newFeeRate.ToString(),
                             oldFeeRate.ToString()));
             }
+        }
+
+        uint64_t nConflictingCount = 0;
+        for (const auto& mi : setIterConflicting) {
+            nConflictingCount += mi->GetCountWithDescendants();
+            // This potentially overestimates the number of actual descendants
+            // but we just want to be conservative to avoid doing too much
+            // work.
+            if (nConflictingCount > MAX_BIP125_REPLACEMENT_CANDIDATES) {
+                return state.Invalid(TxValidationResult::TX_MEMPOOL_POLICY, "too many potential replacements",
+                        strprintf("rejecting replacement %s; too many potential replacements (%d > %d)\n",
+                            hash.ToString(),
+                            nConflictingCount,
+                            MAX_BIP125_REPLACEMENT_CANDIDATES));
+            }
+        }
+        // If not too many to replace, then calculate the set of
+        // transactions that would have to be evicted
+        for (CTxMemPool::txiter it : setIterConflicting) {
+            m_pool.CalculateDescendants(it, allConflicting);
+        }
+        // Check if it's economically rational to mine this transaction rather
+        // than the ones it replaces.
+        for (CTxMemPool::txiter it : allConflicting) {
+            nConflictingFees += it->GetModifiedFee();
+            nConflictingSize += it->GetTxSize();
+        }
 
+        std::set<uint256> setConflictsParents;
+        for (const auto& mi : setIterConflicting) {
```

Comment:
Note, I think this commit is changing a bit the processing, Previously, the 3 operations a) checking candidate feerate is better than any replaced transaction b) fulfilling the set of parents inputs and c) accounting replaced transactions with descendants in the same loop.

After this commit, `setIterConflicting` is iterated 3 times. Overall I think it's okay, though it might add a bit of work in case of well-sized set of replaced transactions. From a DoS viewpoint no changes. An upper bound is still enforced with `nConflictingCount` against `MAX_BIP_125_REPLACEMENT_CANDIDATES` (that would be great to have a set of QA assets with a wide range of mempool candidates payloads to confirm or infirm that kind of intuitions, not sure that's an information you can currently get from `qa-assets/fuzz_seed_corpus/validation_load_mempool`)

Reply:
Right, I was hoping to make this clear in the commit messages. Because of the changes in the preceding commit, we now quit immediately when we see more than `MAX_BIP_125_REPLACEMENT_CANDIDATES`. Previously, we'd grab all potential candidates and only check the total number afterward. So this PR has, overall, reduced the  complexity from O(n) to O(1) where n is the number of conflicts, which is why I think it's okay to iterate 3 times.

The only observable "behavior" change is that we might return `too many potential replacements` instead of `insufficient fee` when a transaction violates both.